sql/src/main/java/io/crate/expression/scalar/StringToArrayFunction.java

```diff
@@ -0,0 +1,162 @@
+        assert args.length == 2 || args.length == 3 : "number of args must be 2 or 3";
+
+        String str = args[0].value();
+        if (str == null) {
+            return null;
+        }
+        if (str.isEmpty()) {
+            return new String[0];
+        }
+
+        String separator = args[1].value();
+        String nullStr = null;
+        if (args.length == 3) {
+            nullStr = args[2].value();
+        }
+
+        return split(str, separator, nullStr);
+    }
+
+    /**
+     * <p>Splits a {@code String} into an array of strings using a separator
+     * string specified and a null-string to set the substring to {@code null}
+     * if they match.
+     * </p>
+     *
+     * <p>
+     * A {@code null} separator splits the string into array in which each
+     * string characters becomes a separate element. If the separator is an
+     * empty string, then the input string is returned as a one element array.
+     * </p>
+     *
+     * @param str       The input {@code String} to split.
+     * @param separator The separator {@code String} used for splitting
+     *                  the input string into substrings. May be {@code null}.
+     * @param nullStr   The null string used to set substrings to {@code null}
+     *                  if they match. May be {@code null}.
+     * @return An array of {@code String}.
+     */
+    private String[] split(
+        @Nonnull String str,
+        @Nullable String separator,
+        @Nullable String nullStr) {
+
+        ArrayList<String> subStrings = new ArrayList<>(str.length());
+
+        if (separator == null) {
+            for (int i = 0; i < str.length(); i++) {
+                String subStr = str.substring(i, i + 1);
+                subStrings.add(setToNullIfMatch.apply(subStr, nullStr));
+            }
+        } else if (separator.isEmpty()) {
+            return new String[]{
+                setToNullIfMatch.apply(str, nullStr)
+            };
+        } else {
+            int start = 0;                      // search start position
+            int pos = str.indexOf(separator);   // separator position
+
+            while (pos >= start) {
+                String subStr;
+
+                if (pos > start) {
+                    subStr = str.substring(start, pos);
+                } else {
+                    // consecutive occurrence of the separator,
+                    // replace it with empty string
+                    subStr = "";
+                }
+
+                start = pos + separator.length();
+                pos = str.indexOf(separator, start);
+
+                subStrings.add(setToNullIfMatch.apply(subStr, nullStr));
+            }
+            String subStr = str.substring(start);
+            subStrings.add(setToNullIfMatch.apply(subStr, nullStr));
+        }
+        return subStrings.toArray(new String[0]);
+    }
+
+    private final BiFunction<String, String, String> setToNullIfMatch =
```

Comment:
What's the advantage of using a `BiFunction` here instead of a regular function with 3 arguments?